packages/vite/src/node/server/hmr.ts

```diff
@@ -250,16 +251,18 @@ function propagateUpdate(
   }
 
   if (node.isSelfAccepting) {
-    boundaries.add({
-      boundary: node,
-      acceptedVia: node,
-    })
+    boundaries.push({ boundary: node, acceptedVia: node })
 
     // additionally check for CSS importers, since a PostCSS plugin like
     // Tailwind JIT may register any file as a dependency to a CSS file.
     for (const importer of node.importers) {
       if (isCSSRequest(importer.url) && !currentChain.includes(importer)) {
-        propagateUpdate(importer, boundaries, currentChain.concat(importer))
+        propagateUpdate(
+          importer,
+          traversedModules,
+          boundaries,
+          currentChain.concat(importer),
```

Comment:
Looks great! Let's merge so we can test it during the beta.

Later on, maybe it is worth exploring using a data structure for `currentChain`/`propagateUpdate`  that doesn't require the creation of new arrays on every call.